File: python/links_client/services/link_db_service.py
"""LinkDBService - Service wrapper for link-cli (clink) database operations"""

import os
import re
import subprocess
from pathlib import Path
from typing import List, Dict, Optional, Any
from links_client.utils.logger import get_logger

logger = get_logger(__name__)

# Default database path
DEFAULT_DB_DIR = Path(__file__).parent.parent.parent.parent / "data"
DEFAULT_DB_FILE = DEFAULT_DB_DIR / "linkdb.links"


class LinkDBService:
    """
    LinkDBService - Wrapper for link-cli database operations
    Uses the clink tool (link-cli) for associative link-based data storage
    """

    def __init__(self, db_path: Optional[str] = None):
        """
        Initialize LinkDBService

        Args:
            db_path: Path to database file (defaults to DEFAULT_DB_FILE)
        """
        self.db_path = db_path or str(DEFAULT_DB_FILE)
        self.next_id = 1  # Track next available ID for menu items

    def execute_query(
        self,
        query: str,
        before: bool = False,
        changes: bool = False,
        after: bool = False,
        trace: bool = False
    ) -> str:
        """
        Execute a clink command

        Args:
            query: LiNo query string
            before: Show state before query
            changes: Show changes
            after: Show state after query
            trace: Show trace information

        Returns:
            Command output

        Raises:
            RuntimeError: If clink is not available or query fails
        """
        flags = []
        if before:
            flags.append('--before')
        if changes:
            flags.append('--changes')
        if after:
            flags.append('--after')
        if trace:
            flags.append('--trace')

        command = ['clink', query, '--db', self.db_path] + flags

        try:
            logger.debug(f"Executing clink command: {' '.join(command)}")

            # Set PATH to include .dotnet/tools directory where clink is installed
            env = os.environ.copy()
            home = os.path.expanduser('~')
            dotnet_tools = os.path.join(home, '.dotnet', 'tools')
            env['PATH'] = f"{dotnet_tools}:{env.get('PATH', '')}"

            result = subprocess.run(
                command,
                capture_output=True,
                text=True,
                env=env,
                check=True
            )

            if result.stderr:
                logger.warning(f"clink command produced stderr: {result.stderr}")

            return result.stdout.strip()

        except FileNotFoundError:
            logger.error("clink command not found - link-cli may not be installed")
            raise RuntimeError(
                "LinkDB not available: clink command not found. "
                "Please install link-cli."
            )
        except subprocess.CalledProcessError as error:
            logger.error(f"Failed to execute clink command: {error}")
            raise RuntimeError(f"LinkDB query failed: {error}")

    def parse_links(self, output: str) -> List[Dict[str, int]]:
        """
        Parse clink output to extract links
        Format: (id: source target)

        Args:
            output: Raw clink output

        Returns:
            List of parsed links
        """
        if not output or not output.strip():
            return []

        lines = [line.strip() for line in output.split('\n') if line.strip()]
        links = []

        for line in lines:
            # Match pattern: (id: source target)
            match = re.match(r'\((\d+):\s+(\d+)\s+(\d+)\)', line)
            if match:
                links.append({
                    'id': int(match.group(1)),
                    'source': int(match.group(2)),
                    'target': int(match.group(3))
                })

        return links

    def create_link(self, source: int, target: int) -> Dict[str, int]:
        """
        Create a new link

        Args:
            source: Source link ID
            target: Target link ID

        Returns:
            Created link

        Raises:
            RuntimeError: If link creation fails
        """
        query = f"() (({source} {target}))"
        output = self.execute_query(query, changes=True)

        # Parse the created link from output
        match = re.search(r'\((\d+):\s+(\d+)\s+(\d+)\)', output)
        if match:
            return {
                'id': int(match.group(1)),
                'source': int(match.group(2)),
                'target': int(match.group(3))
            }

        raise RuntimeError("Failed to parse created link")

    def read_all_links(self) -> List[Dict[str, int]]:
        """
        Read all links from database

        Returns:
            List of all links
        """
        query = "((($i: $s $t)) (($i: $s $t)))"
        output = self.execute_query(query, after=True)
        return self.parse_links(output)

    def read_link(self, link_id: int) -> Optional[Dict[str, int]]:
        """
        Read a specific link by ID

        Args:
            link_id: Link ID

        Returns:
            Link or None if not found
        """
        query = f"((({link_id}: $s $t)) (({link_id}: $s $t)))"
        output = self.execute_query(query, after=True)
        links = self.parse_links(output)
        return links[0] if links else None

    def update_link(self, link_id: int, new_source: int, new_target: int) -> Dict[str, int]:
        """
        Update a link

        Args:
            link_id: Link ID
            new_source: New source value
            new_target: New target value

        Returns:
            Updated link
        """
        query = f"((({link_id}: $s $t)) (({link_id}: {new_source} {new_target})))"
        self.execute_query(query, changes=True)

        return {
            'id': link_id,
            'source': new_source,
            'target': new_target
        }

    def delete_link(self, link_id: int) -> bool:
        """
        Delete a link

        Args:
            link_id: Link ID

        Returns:
            Success status
        """
        query = f"((({link_id}: $s $t)) ())"
        self.execute_query(query, changes=True)
        return True

    def store_menu_item(self, menu_item: Dict[str, Any]) -> int:
        """
        Store menu item data as a link with encoded JSON
        This is a higher-level abstraction for storing menu items

        Args:
            menu_item: Menu item object

        Returns:
            Link ID
        """
        # For now, we'll store the menu item as a link where:
        # - source = incrementing ID
        # - target = menu type identifier

        menu_type_id = 1000  # Identifier for menu items
        item_id = self.next_id
        self.next_id += 1

        link = self.create_link(item_id, menu_type_id)

        # Store the actual JSON data separately (we'll need to implement a string storage mechanism)
        # For now, return the link ID
        return link['id']

    def get_all_menu_items(self) -> List[Dict[str, int]]:
        """
        Get all menu items

        Returns:
            List of menu items
        """
        # Read all links with source pointing to menu type
        all_links = self.read_all_links()

        # Filter links that represent menu items (target = 1000)
        menu_links = [link for link in all_links if link['target'] == 1000]

        return menu_links

    def delete_menu_item(self, link_id: int) -> bool:
        """
        Delete menu item

        Args:
            link_id: Link ID

        Returns:
            Success status
        """
        return self.delete_link(link_id)

    def clear_database(self) -> bool:
        """
        Clear all data from database

        Returns:
            Success status
        """
        try:
            # Delete all links one by one
            all_links = self.read_all_links()
            for link in all_links:
                self.delete_link(link['id'])
            return True
        except Exception as error:
            logger.error(f"Failed to clear database: {error}")
            raise

File: rust/src/services/link_db_service.rs
//! LinkDBService - Service wrapper for link-cli (clink) database operations

use std::env;
use std::path::PathBuf;
use std::process::Stdio;
use tokio::process::Command;
use thiserror::Error;
use tracing::{debug, error, warn};

/// Errors that can occur when using LinkDBService
#[derive(Error, Debug)]
pub enum LinkDBError {
    #[error("LinkDB not available: clink command not found. Please install link-cli.")]
    ClinkNotFound,

    #[error("LinkDB query failed: {0}")]
    QueryFailed(String),

    #[error("Failed to parse link output: {0}")]
    ParseError(String),

    #[error("IO error: {0}")]
    IoError(#[from] std::io::Error),
}

/// A link in the database
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Link {
    /// The unique identifier of the link
    pub id: u64,
    /// The source link ID
    pub source: u64,
    /// The target link ID
    pub target: u64,
}

/// LinkDBService - Wrapper for link-cli database operations
/// Uses the clink tool (link-cli) for associative link-based data storage
pub struct LinkDBService {
    db_path: PathBuf,
    next_id: u64,
}

impl LinkDBService {
    /// Create a new LinkDBService with the specified database path
    ///
    /// # Arguments
    ///
    /// * `db_path` - Optional path to the database file. If None, uses the default path.
    pub fn new(db_path: Option<PathBuf>) -> Self {
        let default_path = PathBuf::from("data/linkdb.links");
        Self {
            db_path: db_path.unwrap_or(default_path),
            next_id: 1,
        }
    }

    /// Execute a clink command
    ///
    /// # Arguments
    ///
    /// * `query` - LiNo query string
    /// * `before` - Show state before query
    /// * `changes` - Show changes
    /// * `after` - Show state after query
    /// * `trace` - Show trace information
    pub async fn execute_query(
        &self,
        query: &str,
        before: bool,
        changes: bool,
        after: bool,
        trace: bool,
    ) -> Result<String, LinkDBError> {
        let mut args = vec![query.to_string(), "--db".to_string(), self.db_path.to_string_lossy().to_string()];

        if before {
            args.push("--before".to_string());
        }
        if changes {
            args.push("--changes".to_string());
        }
        if after {
            args.push("--after".to_string());
        }
        if trace {
            args.push("--trace".to_string());
        }

        debug!("Executing clink command: clink {}", args.join(" "));

        // Set PATH to include .dotnet/tools directory where clink is installed
        let home = env::var("HOME").unwrap_or_else(|_| String::new());
        let dotnet_tools = format!("{}/.dotnet/tools", home);
        let path = format!("{}:{}", dotnet_tools, env::var("PATH").unwrap_or_default());

        let output = Command::new("clink")
            .args(&args)
            .env("PATH", &path)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .output()
            .await;

        match output {
            Ok(output) => {
                if !output.status.success() {
                    let stderr = String::from_utf8_lossy(&output.stderr);
                    if stderr.contains("not found") || stderr.contains("command not found") {
                        error!("clink command not found - link-cli may not be installed");
                        return Err(LinkDBError::ClinkNotFound);
                    }
                    error!("Failed to execute clink command: {}", stderr);
                    return Err(LinkDBError::QueryFailed(stderr.to_string()));
                }

                let stderr = String::from_utf8_lossy(&output.stderr);
                if !stderr.is_empty() {
                    warn!("clink command produced stderr: {}", stderr);
                }

                Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
            }
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
                error!("clink command not found - link-cli may not be installed");
                Err(LinkDBError::ClinkNotFound)
            }
            Err(e) => {
                error!("Failed to execute clink command: {}", e);
                Err(LinkDBError::IoError(e))
            }
        }
    }

    /// Parse clink output to extract links
    /// Format: (id: source target)
    pub fn parse_links(&self, output: &str) -> Vec<Link> {
        if output.trim().is_empty() {
            return Vec::new();
        }

        let mut links = Vec::new();
        let re = regex_lite::Regex::new(r"\((\d+):\s+(\d+)\s+(\d+)\)").unwrap();

        for line in output.lines() {
            if let Some(caps) = re.captures(line.trim()) {
                if let (Some(id), Some(source), Some(target)) =
                    (caps.get(1), caps.get(2), caps.get(3))
                {
                    if let (Ok(id), Ok(source), Ok(target)) = (
                        id.as_str().parse::<u64>(),
                        source.as_str().parse::<u64>(),
                        target.as_str().parse::<u64>(),
                    ) {
                        links.push(Link { id, source, target });
                    }
                }
            }
        }

        links
    }

    /// Create a new link
    ///
    /// # Arguments
    ///
    /// * `source` - Source link ID
    /// * `target` - Target link ID
    pub async fn create_link(&self, source: u64, target: u64) -> Result<Link, LinkDBError> {
        let query = format!("() (({} {}))", source, target);
        let output = self.execute_query(&query, false, true, false, false).await?;

        let re = regex_lite::Regex::new(r"\((\d+):\s+(\d+)\s+(\d+)\)").unwrap();
        if let Some(caps) = re.captures(&output) {
            if let (Some(id), Some(src), Some(tgt)) = (caps.get(1), caps.get(2), caps.get(3)) {
                if let (Ok(id), Ok(source), Ok(target)) = (
                    id.as_str().parse::<u64>(),
                    src.as_str().parse::<u64>(),
                    tgt.as_str().parse::<u64>(),
                ) {
                    return Ok(Link { id, source, target });
                }
            }
        }

        Err(LinkDBError::ParseError("Failed to parse created link".to_string()))
    }

    /// Read all links from database
    pub async fn read_all_links(&self) -> Result<Vec<Link>, LinkDBError> {
        let query = "((($i: $s $t)) (($i: $s $t)))";
        let output = self.execute_query(query, false, false, true, false).await?;
        Ok(self.parse_links(&output))
    }

    /// Read a specific link by ID
    pub async fn read_link(&self, id: u64) -> Result<Option<Link>, LinkDBError> {
        let query = format!("((({}:$s $t)) (({}:$s $t)))", id, id);
        let output = self.execute_query(&query, false, false, true, false).await?;
        let links = self.parse_links(&output);
        Ok(links.into_iter().next())
    }

    /// Update a link
    ///
    /// # Arguments
    ///
    /// * `id` - Link ID
    /// * `new_source` - New source value
    /// * `new_target` - New target value
    pub async fn update_link(
        &self,
        id: u64,
        new_source: u64,
        new_target: u64,
    ) -> Result<Link, LinkDBError> {
        let query = format!("((({}:$s $t)) (({}: {} {})))", id, id, new_source, new_target);
        self.execute_query(&query, false, true, false, false).await?;

        Ok(Link {
            id,
            source: new_source,
            target: new_target,
        })
    }

    /// Delete a link
    pub async fn delete_link(&self, id: u64) -> Result<bool, LinkDBError> {
        let query = format!("((({}:$s $t)) ())", id);
        self.execute_query(&query, false, true, false, false).await?;
        Ok(true)
    }

    /// Store menu item data as a link
    /// This is a higher-level abstraction for storing menu items
    pub async fn store_menu_item(&mut self) -> Result<u64, LinkDBError> {
        let menu_type_id = 1000u64; // Identifier for menu items
        let item_id = self.next_id;
        self.next_id += 1;

        let link = self.create_link(item_id, menu_type_id).await?;
        Ok(link.id)
    }

    /// Get all menu items
    pub async fn get_all_menu_items(&self) -> Result<Vec<Link>, LinkDBError> {
        let all_links = self.read_all_links().await?;
        Ok(all_links.into_iter().filter(|link| link.target == 1000).collect())
    }

    /// Delete menu item
    pub async fn delete_menu_item(&self, link_id: u64) -> Result<bool, LinkDBError> {
        self.delete_link(link_id).await
    }

    /// Clear all data from database
    pub async fn clear_database(&self) -> Result<bool, LinkDBError> {
        let all_links = self.read_all_links().await?;
        for link in all_links {
            self.delete_link(link.id).await?;
        }
        Ok(true)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_links() {
        let service = LinkDBService::new(None);

        let output = "(1: 2 3)\n(4: 5 6)";
        let links = service.parse_links(output);

        assert_eq!(links.len(), 2);
        assert_eq!(links[0], Link { id: 1, source: 2, target: 3 });
        assert_eq!(links[1], Link { id: 4, source: 5, target: 6 });
    }

    #[test]
    fn test_parse_empty_output() {
        let service = LinkDBService::new(None);

        let output = "";
        let links = service.parse_links(output);

        assert!(links.is_empty());
    }
}
